### DISCUSSION TITLE
동시성 제어 처리 - 판매자 우선 변경 issue

### DISCUSSION THREAD

**Discussion by YurimYang** [SCBJ-7/SCBJ-BE#11] [Q&A] [answered]
### 안건

**예제1)** 
- 1. 판매자 : 22만원 -> 23만원 변경 시도
- 2. 구매자 : 22만원을 read해서 트랜잭션 시작
- 3. 판매자 : 22 -> 23만원 변경 완료
- 4. 구매자 : ??? 22만원으로 읽었는데 23만원으로 변경되서 결제 실패
![동시성 처리 1번 문제](https://github.com/SCBJ-7/SCBJ-BE/assets/63856521/267e29c4-b754-4d03-bc36-d7301d80561a)

### 고민사항

-  구매자 입장에서는 결제가 실패되어야한다.
- 다만, 동시성을 고려했을 때, 결제를 성공 -> 환불 로 해야할지, 바로 실패로 해야할지 고민이 된다.

case 1) 결제 성공 -> 환불 
--
- 판매자 금액 변경 시점 기록, 구매자 상품 구매 시점 기록
- 구매자 상품 구매 시점(15시 05분) > 판매자 금액 변경 시점(15시 03분) : 구매자 성공 -> 환불
- 단, 환불 시점을 언제로 해야할지 의문이 든다.

case 2 ) 실패
--
- 트랜잭션이 이어지는 거랑 별 다를 것이 없기 때문이 불가능하다고 생각이 든다.

**Comment by Aleexender** [comment, marked as answer]
## 레디스 분산 락 - 확장성

**레디스 특징**

- key-value 형식의 데이터 저장소
- 단일 스레드 실행을 하기 때문에 Atomic 하다 ( 한 트랜잭션의 연산들이 모두 성공 혹은 실패)
- 디스크가 아닌 인메모리에 저장(휘발성)

여러가지 사용처

cahce, pipelining, key space 등 여러가지 사용법이 있음

**분산락**

분산락이란 서로 다른 프로세스에서 동일한 공유자원에 접근할 때, 데이터의 원자성을 보장하기위해 활용되는 방법. 

Redlock 이라는 알고리즘을 제안

We propose an algorithm, called **Redlock**, which implements a DLM(데이터를 입력할때부터 삭제할 때까지 라이프사이클의 전 범위에서 데이터를 관리하는 접근방식) which we believe to be safer than the vanilla single instance approach. We hope that the community will analyze it, provide feedback, and use it as a starting point for the implementations or more complex or alternative designs.

N 인스턴에서 순차적으로 잠금을 획득하려고 시도함

In the distributed version of the algorithm we assume we have N Redis masters. Those nodes are totally independent, so we don’t use replication or any other implicit coordination system. We already described how to acquire and release the lock safely in a single instance. We take for granted that the algorithm will use this method to acquire and release the lock in a single instance. In our examples we set N=5, which is a reasonable value, so we need to run 5 Redis masters on different computers or virtual machines in order to ensure that they’ll fail in a mostly independent way.

In order to acquire the lock, the client performs the following operations:

It gets the current time in milliseconds.
It tries to acquire the lock in all the N instances sequentially, using the same key name and random value in all the instances. During step 2, when setting the lock in each instance, the client uses a timeout which is small compared to the total lock auto-release time in order to acquire it. For example if the auto-release time is 10 seconds, the timeout could be in the ~ 5-50 milliseconds range. This prevents the client from remaining blocked for a long time trying to talk with a Redis node which is down: if an instance is not available, we should try to talk with the next instance ASAP.
The client computes how much time elapsed in order to acquire the lock, by subtracting from the current time the timestamp obtained in step 1. If and only if the client was able to acquire the lock in the majority of the instances (at least 3), and the total time elapsed to acquire the lock is less than lock validity time, the lock is considered to be acquired.
If the lock was acquired, its validity time is considered to be the initial validity time minus the time elapsed, as computed in step 3.
If the client failed to acquire the lock for some reason (either it was not able to lock N/2+1 instances or the validity time is negative), it will try to unlock all the instances (even the instances it believed it was not able to lock).
Is the Algorithm Asynchronous?
The algorithm relies on the assumption that while there is no synchronized clock across the processes, the local time in every process updates at approximately at the same rate, with a small margin of error compared to the auto-release time of the lock. This assumption closely resembles a real-world computer: every computer has a local clock and we can usually rely on different computers to have a clock drift which is small.

At this point we need to better specify our mutual exclusion rule: it is guaranteed only as long as the client holding the lock terminates its work within the lock validity time (as obtained in step 3), minus some time (just a few milliseconds in order to compensate for clock drift between processes).

This paper contains more information about similar systems requiring a bound clock drift: Leases: an efficient fault-tolerant mechanism for distributed file cache consistency.

## 레디슨 - 자바에서 쓸수있는 클라이언트

Spring에서 제공하는 대표적인 redis 라이브러리로는 Lettuce는 다음과 같은 존재

1. Lock 의 타임아웃이 지정되있지 않음
    1. 락 획득 후 모종의 이유로 어플리케이션이 종료된다거나 했을 경우엔 락은 해제되지 않음 그렇게 되면 타 프로세스들은 락이 해제되기만을 기다리는 무한정 대기상태로 빠지게 됨, 이는 곧 시스템 장애로 이어짐 → spin lock 때문 → Lock을 획득하려는 스레드가 Lock을 획득하려는 스레드가 Lock을 획들할 수 있는지 확인하면서 반복적으로 시도하는 방법

![Untitled](https://prod-files-secure.s3.us-west-2.amazonaws.com/1acacc8f-fc15-45a0-ad09-4ec6a61b1fd2/10d1cfa3-7493-4b42-9fd7-10f701fd4e5d/Untitled.png)

**Redission 은 Lettuce에 비해 다음과 같은 장점이 있음**

1. Lock에 타임아웃을 지정할 수 있음
    1. Redisson은 락 획득시도시 타임아웃을 명시하게 되어있음 → 무한대기 상태 빠질 가능성 X
2. pub/sub 방식을 사용하므로 스핀락을 사용하지 않음
    1. 락이 해제되면 락을 subscribe 하는 클라이언트들은 더 이상 락을 획득해도 되냐고 redis로 요청 보내지 않음

**Pub/sub**

MSA 아키텍쳐 안에 있는 어싱크 커뮤니케이팅

pub/sub 을 사용하려면 4가지 중요한 컨셉이 있음

Message: 서로 다른 당사자 간에 교환되도록 인코딩된 직렬화된 개별 통신 단위 

Publisher: 메시지 전송을 담당하는 어플맄이션 또는 기타 엔티티

SubScribrer: 하나 이상의 게시자로부터 메시지를 수신하는 어플리케이션 또는 기타  엔티티 

Topic: 특정 주제에 대한 메시지를 포함하는 채널

가장 중요한건 발신자나 수신자 모두 상대방의 존재를 알필요 없다는 것. 즉 둘다 독립적으로 작동할 수 있다.

Publisher는 누가 메시지를 구독을 위하 받는지 알필요 없다. subscriber는 누가 메시지를 보내는지 알필요 없다. 

어플리케이션이 특정 주제에 관심이 있는 경우 해당 주제를 구독하기만 하면 즉시 메시지 수신을 시작함. 이는 수신자가 메시지를 검색할때까지 메시지 일괄 처리하는 메시지 대기열의 개념과 다릅니다.

**Pub/sub vs observer 패턴**

보통 observer 패턴이랑 비교 되고 대조된다

observer 패턴에서 주체 또는 관찰 기능으로 알려진 단일 개체는 observable 라고 알려진 종속 항목 목록을 유지 관리하는 일을 담당함. 주체의 상태가 변경되면 모든 종속 observers 들이 자동으로 업데이트 된다.

효율성: pub / sub 패턴을 사용하면 리스너들이 새 메시지를 반복적으로 poll할 필요가 없음. 대신 메시지가 즉시 전달되는데 이는 실시간 어플리케이션의 주요 이점임. 

확장성: 퍼블리셔와 subscriber를 분리하여 각 어플리케이션이 독립적으로 작동할수 있음. 이를 통해 대규모 시스템을 더 쉽게 확장하고 시스템의 일부만 변경할수 있음

단순성: 각 어플리케이션은 잠재적인 실패 지점을 나타냄. 개별 어플리케이션 간의 연결을 제거하고 이를 다른 주제에 대한 연결로 대체하여 이러한 복잡성을 상당 부분 제거.

https://www.baeldung.com/redis-redisson

### 아이솔레이션 레벨 -정책(보안수준)

동시에 여러 트랜잭션이 처리될 때

특정 트랜잭션이 다른 트랜잭션에서 변경하거나 조회하는 데이터를 볼수 있도록 할지말지 결정

![IMG_0105.GIF](https://s3-us-west-2.amazonaws.com/secure.notion-static.com/40b47f71-0211-49cf-8741-2b3e70a5f2c7/IMG_0105.gif)

Read Uncommited - 어떤 트랜잭션의 변경내용이 commit이나 rollback과 상관없이 다른 트랜잭션에서 보여진다

Read Committed - 어떤 트랜잭션의 변경 내용이 commit 되어야만 다른 트랜잭션에서 조회 가능(온라인 서비스에서 가장 많이 선택됨)

Repeatable Read - 디폴트 level(MySQL), 트렌잭션이 시작되기 전에 커밋된 내용에 대해서만 조회할 수있는

격리수준, NON -REPETABLE DEAD 부적합이 발생 X

Consistent Read - read(=select) opertaion을 수행할때 현재 DB의 값이 아닌 특정 시점의 DB snapshot을 읽어옴

Serializable - 가장 단순하고 엄격한 격리수준 InnoDB에서 기본적으로 순수한 SELECT 작업은 아무런 잠금을 걸지않고 동작, SERIALIZABLE일 경우 읽기 작업에도 공유잠금을 설정, 다른트랜잭션에서 이 레코드를 변경하지 못하게 됨, 동시처리 능력이 다른 격리수준보다 떨어짐 + 성능저하 발생

## 스프링 프로파게이션

Spring 이 제공 하는 선언적 트랜잭션의 장점중 하나는 여러 트랜잭션을 묶어서 커다란 하나의 트랜잭션 경계를 만들수 있음. 작업하다 보면 기존에 트랜잭션이 진행중일 때 추가적인 트랜잭션을 진행하는 경우가 있음.

이미 트랜잭션이 진행중일때 추가 트랜잭션을 진행을 어떻게 할지 결정하는 것이 전파 속성(Propagation)이다.

전파 속성에 따라 기존의 트랜잭션에 참여할 수도 있고, 별도의 트랜잭션으로 진행할 수도 있고, 에러를 발생시키는 등 여러 선택을 할 수 있음. 

**물리 트랜잭션과 논리 트랜잭션**

트랜잭션은 데이터베이스에서 제공하는 기술이므로 커넥션 객체를 통해 처리함. 그래서 1개의 트랜잭션을 사용한다는 것은 하나의 커넥션 객체를 사용한다는 것, 실제 데이터베이스의 트랜잭션을 사용한다는 점에서 **물리 트랜잭션이라고 함.** 

![Untitled](https://prod-files-secure.s3.us-west-2.amazonaws.com/1acacc8f-fc15-45a0-ad09-4ec6a61b1fd2/11cd3ae8-1e83-4ec7-9993-86571a7b9da5/Untitled.png)

물리 트랜잭션: 실제 데이터베이스에 적용되는 트랜잭션으로, 커넥션을 통해 커밋/롤백하는 단위

논리 트랜잭션: 스프링이 트랜잭션 매니저를 통해 트랜잭션을 처리하는 단위

모든 논리 트랜잭션이 커밋되어야 물리 트랜잭션이 커밋됨

하나의 논리 트랜잭션이라도 롤백되면 물리 트랜잭션은 롤백됨

**REQUIRED**

 부모 트랜잭션 내에서 실행하며 부모 트랜잭션이 없을 경우 새로운 트랜잭션을 생성

의미: 트랜잭션이 필요함(없으면 새로 만듬)

기존 트랜잭션 없음: 새로운 트랜잭션을 생성

기존 트랜잭션이 있음: 기존 트랜잭션에 참여

> REQUIRED는 디폴트 속성으로써 모든 트렌젝션 매니저가 지원하는 속성. 별도의 설정이 없다면  REQUIRED로 트랜잭션이 진행
> 

**SUPPORTS**

의미: 트랜잭션이 있으면 지원함(트랜잭션이 없어도 됨)

기존트랜잭션 없음: 트랜잭션 없이 진행함

기존 트랜잭션이 있음: 기존 트랜잭션에 참여함

**MANDATORY**

의미: 트랜잭션이 의무임(트랜잭션이 반드시 필요함)

기존 트랜잭션 없음: IllegalTransactionStateException 예외 발생

기존 트랜잭션이 있음: 기존 트랜잭션에 참여함

**REQUIRES_NEW**

의미: 항상 새로운 트랜잭션이 필요함

기존 트랜잭션 없음: 새로운 트랜잭션을 생성함

기존 트랜잭션이 있음: 기존 트랜잭션을 보류시키고 새로운 트랜잭션을 생성함

**NOT_SUPPORTED**

의미: 트랜잭션을 지원하지 않음(트랜잭션 없이 진행)

기존 트랜잭션 없음: 트랜잭션 없이 진행함

기존 트랜잭션이 있음: 기존 트랜잭션을 보류시키고 트랜잭션 없이 진행함

**NEVER**

의미: 트랜잭션을 사용하지 않음(기존 트랜잭션도 허용하지 않음)

기존 트랜잭션 없음: 트랜잭션 없이 진행 

기존 트랜잭션이 있음: IllegalTransactionException 예외 발생

**NESTED**

의미: 중첩(자식) 트랜잭션을 생성

기존 트랜잭션 없음: 새로운 트랜잭션을 생성

기존 트랜잭션 있음: 중첩 트랜잭션을 만듬

![Untitled](https://prod-files-secure.s3.us-west-2.amazonaws.com/1acacc8f-fc15-45a0-ad09-4ec6a61b1fd2/24b7e48a-24ef-4086-8c91-730736eb9900/Untitled.png)

https://mangkyu.tistory.com/269

![Untitled](https://prod-files-secure.s3.us-west-2.amazonaws.com/1acacc8f-fc15-45a0-ad09-4ec6a61b1fd2/cb35c53c-1054-40f4-bab4-18d04c7d0d2a/Untitled.png)

**팬텀 리드**

데이터를 읽은 후 로직을 진행하는데 있어서, 다른 트랜잭션의 UPDATE, INSERT 의 개입으로 다시 읽었을때 데이터의 무결성이 깨진다는 의미

https://catch-me-java.tistory.com/60

https://techblog.woowahan.com/2631/ - mysql 로 분산락 구현

**Reply by Aleexender** [reply]
Redis 를 사용하게 되면 비용도 고려를 해야하는데 
맨마지막에 제공하는 사이트(우테코) 에서 mySQL로 분산락을 구현해볼수도 있네요!
이렇게 되면 비용을 따로 청구하지 않고도 분산락 구현이 가능할거같네요!

**Reply by YurimYang** [reply]
궁금한 점이 있습니다 ! 분산락 자체가 여러 서버가 동시에 자원에 접근하는 상황에서 (Redis를 활용해서) 하나의 서버만 접근 가능하도록 하는 기술이라고 알고 있습니다. 그렇다면 저희의 서비스에서는 구매자 서버와 판매자 서버를 다르게 구현해서 분산락을 구현하자고 언급하신걸까요 ?

**Comment by wocjf0513** [comment]
공부를 해도 온전히 안 다고 할 수 없어 -> 이견이 있으면 알려준다는 마음으로 대댓글 남겨주세요!

[해결 방안]
1. 정훈님이 말하신 namedLock을 이용
=> mysql을 사용하기 때문에 비용적인 측면에서 이득을 볼 수 있고, 저희가 이 전에 mysql 사용하기로 결정했기 때문에
더더욱 괜찮은 거 같습니다.
다만, mysql을 주로 쓰는 우리 서비스에서, 동시성 처리를 위해서 또, mysql을 쓰게 되면 mysql db의 부하가 심해져서 성능 상 mysql 과 redis를 나눠 사용하는 것보다 느릴 수 있다고 생각합니다.

2. redis lock을 이용
=> 2는 비용적인 측면을 고려하자면, 좋지는 않지만 앞으로 msa구조의 아키텍쳐를 사용하고 경험하기 전에, 이를 통해 구현하는 것도 나쁘지 않다고 생각합니다.

3. 구글에 '보상트랜젝션'이라고 올라와 있는 것(msa상에서)과는 조금 다른 보상 트랜젝션을 사용할 수 있을 거 같습니다.
=>보통 보상트랜젝션이 msa 구조에서 카프카의 메시징 큐를 이용해 실패한 트랜젝션에 대한 다중 서버 db의 데이터 정합성을 지키기 위해서 사용되는 거 같습니다.
결제 트랜젝션이 끝난 이후, aop를 이용해 현재 양도 글의 가격과 구매 내역의 가격을 비교해 보상 트랜젝션을 작성하면, 결제 트랜젝션이 commit된 이후에 검증을 하게 될 수 있습니다. !

4. 정책 상 가격 변경 한 이후에 ~ 2초 동안의 이전 가격 결제 내역이 발생할 수 있습니다.
이런 식으로 뺄 수 있을 거 같습니다.
Transaction 결제 이루어지는 시간 (1초)+ FCM을 통해 푸쉬 알림을 주는 시간 (보통 1초라고 합니다.) => 많이 잡아서 2~3 초가 될 거 같은데 가격 변동 표시를 보고 2~3초 그렇게 사용자가 불편함을 느낄 고민이 되긴 합니다.

**Reply by YurimYang** [reply]
여러 블로그를 찾아보면서 현재 우아한형제들에서도 mySQL을 활용해서 분산Lock을 구현했다고 합니다 
한번 참고해보시는 거 추천드려요 ~!! 
🔗 https://techblog.woowahan.com/2631/

**Reply by wocjf0513** [reply]
감사합니다.

**Comment by YurimYang** [comment]
저는 ISOLATION LEVEL에 따라 동시성 제어하는 방법에 대해 공부해봤습니다.
<--공부한 내용 링크로 걸어둡니다!-->
🔗 https://www.notion.so/dev-if/25fac09ae75b4414920f6f4d3e6f7b43#097fddb795ae4218a9bfe325580110b6
공부하면서 느낀 점이 
- 저희가 MySQL을 사용하기로 했고, 
- MySQL은 Gap Lock(잠금대기)를 기본적으로 제공해주기에 Phantom Read가 발생하지 않는다
- 위의 2점을 고려해서 Repeatable Read한 상황으로 구현하면 case1이 해결된다고 생각합니다! 
추가적으로 namedLock에 대해서도 공부해서 가겠습니다 ~!!!

**Comment by Kwonminwoo** [comment]
해당 문제를 어떻게 해결할 수 있을까에 대해 고민하고 공부하였습니다.
구매자가 구매 버튼을 눌렀을 때 해당 상품에 lock을 걸어 다른 스레드에서 접근하지 못하도록 하는 방법입니다. 구매자가 구매 페이지에 접근한 순간부터 락이 걸리므로 판매자가 가격 변경 확인 버튼을 눌렀을 때 해당 데이터가 변경되는 것을 막을 수 있을 것 같습니다.
이 때 락을 거는 방법은 mySql에 row lock을 거는 방법과 레디스를 이용한 방법이 있을 것 같습니다.  redis를 이용한 방법은 정훈님이 말씀하신대로 부가적인 공수가 들겠지만 나아가 분산 환경을 고려 했을 때, 혹은 캐싱과 같은 작업에도 레디스를 활용하면 레디스를 쓸 때 실보다 득이 많을 수도 있을 것 같습니다.
어떤 방법을 이용할지에 대해서는 더 얘기해보면 좋을 것 같습니다.

**Comment by YurimYang** [comment]
### **<--최종결과-->**

1. Redis Redisson을 활용하여 분산락 사용
2. MySQL로 naemdLock 활용

**Reply by wocjf0513** [reply]
[Redisson 사용 시 참고 기술 블로그]
https://helloworld.kurly.com/blog/distributed-redisson-lock/
https://techblog.woowahan.com/2709/
